File: ui/linux/src/ui/EditingSurface.cpp
#include "EditingSurface.hpp"

#include "model/GridEditing.hpp"

#include <QDialogButtonBox>
#include <QFrame>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QScrollArea>
#include <QStringList>
#include <QVBoxLayout>

namespace {

QLabel* wrappedLabel(const QString& text, QWidget* parent) {
    auto* label = new QLabel(text, parent);
    label->setWordWrap(true);
    label->setTextFormat(Qt::PlainText);
    return label;
}

}  // namespace

// --- the bar ----------------------------------------------------------------

StagedEditsBar::StagedEditsBar(dg::PendingEdits* edits, QWidget* parent)
    : QWidget(parent), edits_(edits) {
    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(12, 6, 12, 6);
    layout->setSpacing(10);

    headline_ = new QLabel(this);
    QFont bold = headline_->font();
    bold.setBold(true);
    headline_->setFont(bold);
    detail_ = new QLabel(this);
    detail_->setTextFormat(Qt::PlainText);

    auto* text = new QVBoxLayout();
    text->setContentsMargins(0, 0, 0, 0);
    text->setSpacing(1);
    text->addWidget(headline_);
    text->addWidget(detail_);
    layout->addLayout(text, 1);

    discardButton_ = new QPushButton(QStringLiteral("Discard"), this);
    resolveButton_ = new QPushButton(this);
    commitButton_ = new QPushButton(this);
    reloadButton_ = new QPushButton(QStringLiteral("Reload"), this);
    connect(discardButton_, &QPushButton::clicked, this,
            &StagedEditsBar::discardRequested);
    connect(resolveButton_, &QPushButton::clicked, this,
            &StagedEditsBar::resolveRequested);
    connect(commitButton_, &QPushButton::clicked, this,
            &StagedEditsBar::commitRequested);
    connect(reloadButton_, &QPushButton::clicked, this,
            &StagedEditsBar::reloadRequested);
    layout->addWidget(discardButton_);
    layout->addWidget(resolveButton_);
    layout->addWidget(commitButton_);
    layout->addWidget(reloadButton_);

    setAutoFillBackground(true);
    connect(edits_, &dg::PendingEdits::stagingChanged, this,
            &StagedEditsBar::refresh);
    refresh();
}

void StagedEditsBar::setCommitting(bool committing) {
    committing_ = committing;
    refresh();
}

void StagedEditsBar::setRereading(bool rereading) {
    rereading_ = rereading;
    refresh();
}

void StagedEditsBar::refresh() {
    if (edits_->isEmpty()) {
        hide();
        return;
    }
    const int pending = edits_->pendingCount();
    const int written = edits_->documents().size() - pending;
    const int conflicts = edits_->conflictCount();

    if (pending == 0) {
        // Everything staged was written. The grid still shows the rows as they
        // were loaded with the typed values drawn over them, so the only way
        // to see what the server holds is to ask it again — which is what the
        // reload button offers.
        headline_->setText(
            written == 1
                ? QStringLiteral("1 document written — the grid still shows what was loaded")
                : QStringLiteral("%1 documents written — the grid still shows what was loaded")
                      .arg(written));
    } else {
        headline_->setText(
            pending == 1
                ? QStringLiteral("1 document edited, not yet written")
                : QStringLiteral("%1 documents edited, not yet written").arg(pending));
    }

    // Updates and deletes are different enough that a single count would hide
    // one behind the other.
    QStringList parts;
    if (edits_->updateCount() > 0) {
        parts << QStringLiteral("%1 to update").arg(edits_->updateCount());
    }
    if (edits_->deleteCount() > 0) {
        parts << QStringLiteral("%1 to delete").arg(edits_->deleteCount());
    }
    if (written > 0) {
        parts << QStringLiteral("%1 already written").arg(written);
    }
    if (conflicts > 0) {
        parts << (conflicts == 1
                      ? QStringLiteral("1 changed on the server")
                      : QStringLiteral("%1 changed on the server").arg(conflicts));
    }
    detail_->setText(parts.join(QStringLiteral(" · ")));
    detail_->setVisible(!parts.isEmpty());

    if (committing_) {
        headline_->setText(QStringLiteral("committing…"));
        discardButton_->hide();
        resolveButton_->hide();
        commitButton_->hide();
        reloadButton_->hide();
    } else if (pending == 0) {
        discardButton_->hide();
        resolveButton_->hide();
        commitButton_->hide();
        reloadButton_->show();
    } else {
        discardButton_->show();
        // A conflicted edit cannot simply be committed again — the same guard
        // would refuse it — so the way forward is offered where the refusal is
        // visible, not only inside the report.
        resolveButton_->setVisible(conflicts > 0);
        resolveButton_->setText(
            conflicts == 1 ? QStringLiteral("Resolve 1 Conflict…")
                           : QStringLiteral("Resolve %1 Conflicts…").arg(conflicts));
        resolveButton_->setEnabled(!rereading_);
        commitButton_->show();
        commitButton_->setText(pending == 1
                                   ? QStringLiteral("Commit 1…")
                                   : QStringLiteral("Commit %1…").arg(pending));
        reloadButton_->hide();
    }
    show();
}

// --- the report -------------------------------------------------------------

MutationReportDialog::MutationReportDialog(const dg::MutationReport& report,
                                           QWidget* parent)
    : QDialog(parent) {
    setWindowTitle(QStringLiteral("Commit report"));
    resize(560, 420);

    auto* layout = new QVBoxLayout(this);

    const QString title =
        report.isClean()
            ? (report.applied == 1
                   ? QStringLiteral("1 document written")
                   : QStringLiteral("%1 documents written").arg(report.applied))
            : QStringLiteral("The batch stopped part way through");
    auto* titleLabel = new QLabel(title, this);
    QFont bold = titleLabel->font();
    bold.setBold(true);
    titleLabel->setFont(bold);
    layout->addWidget(titleLabel);

    QString subtitle = QStringLiteral("%1 applied").arg(report.applied);
    if (report.failed > 0) {
        subtitle += QStringLiteral(" · %1 failed").arg(report.failed);
    }
    if (report.notAttempted > 0) {
        subtitle += QStringLiteral(" · %1 never attempted").arg(report.notAttempted);
        subtitle += QStringLiteral(
            ". The ones that were never attempted are still staged — nothing was "
            "written for them, and nothing was lost.");
    }
    if (report.conflicts > 0) {
        subtitle += QStringLiteral(
            " A version conflict means the document changed on the server after "
            "you loaded it, so the write was refused rather than overwriting "
            "someone else's change. What you typed is still staged — resolve it "
            "to see what changed.");
    }
    layout->addWidget(wrappedLabel(subtitle, this));

    auto* list = new QWidget(this);
    auto* listLayout = new QVBoxLayout(list);
    listLayout->setContentsMargins(4, 4, 4, 4);
    listLayout->setSpacing(8);

    for (const dg::MutationNotice& notice : report.notices) {
        const QString prefix = notice.isWarning() ? QStringLiteral("⚠ ")
                                                  : QStringLiteral("ⓘ ");
        QString text = prefix + notice.message;
        if (!notice.code.isEmpty()) {
            text += QStringLiteral("  [%1]").arg(notice.code);
        }
        listLayout->addWidget(wrappedLabel(text, list));
    }
    if (!report.notices.isEmpty()) {
        auto* rule = new QFrame(list);
        rule->setFrameShape(QFrame::HLine);
        listLayout->addWidget(rule);
    }

    for (const dg::MutationRow& row : report.rows) {
        QString mark;
        QString detail;
        switch (row.outcome) {
            case dg::MutationRow::Outcome::Applied: {
                mark = QStringLiteral("✓");
                detail = QStringLiteral("written");
                if (row.seqNo) {
                    detail += QStringLiteral(" · now at _seq_no %1").arg(*row.seqNo);
                }
                if (row.forcedRefresh) {
                    detail += QStringLiteral(
                        " · the server forced an immediate refresh rather than "
                        "waiting for one");
                }
                break;
            }
            case dg::MutationRow::Outcome::NotAttempted:
                mark = QStringLiteral("…");
                detail = QStringLiteral(
                    "never attempted — the batch stopped before it, so this is "
                    "still staged");
                break;
            case dg::MutationRow::Outcome::Failed:
                mark = row.conflict ? QStringLiteral("⑂") : QStringLiteral("✗");
                detail = row.conflict
                             ? QStringLiteral(
                                   "version conflict — this document changed on the "
                                   "server after you loaded it, so nothing was written")
                             : (row.error.isEmpty()
                                    ? QStringLiteral("the write failed")
                                    : row.error);
                break;
        }
        listLayout->addWidget(wrappedLabel(
            QStringLiteral("%1  %2  %3/%4\n     %5")
                .arg(mark, row.op, row.index, row.documentId, detail),
            list));
    }
    listLayout->addStretch(1);

    auto* scroll = new QScrollArea(this);
    scroll->setWidgetResizable(true);
    scroll->setWidget(list);
    layout->addWidget(scroll, 1);

    const QString footer =
        report.conflicts > 0
            ? QStringLiteral("Reads each conflicted document back and shows what changed.")
            : (report.isClean()
                   ? QString()
                   : QStringLiteral("Re-run the statement to see what the server holds now."));
    if (!footer.isEmpty()) {
        layout->addWidget(wrappedLabel(footer, this));
    }

    auto* buttons = new QDialogButtonBox(this);
    if (report.conflicts > 0) {
        QPushButton* resolve = buttons->addButton(
            QStringLiteral("Resolve Conflicts…"), QDialogButtonBox::ActionRole);
        connect(resolve, &QPushButton::clicked, this, [this]() {
            emit resolveConflictsRequested();
            accept();
        });
    }
    QPushButton* done =
        buttons->addButton(QStringLiteral("Done"), QDialogButtonBox::AcceptRole);
    done->setDefault(true);
    connect(buttons, &QDialogButtonBox::accepted, this, &QDialog::accept);
    layout->addWidget(buttons);
}

File: ui/macos/Sources/datagrep-app/EditingSurface.swift
import DatagrepKit
import SwiftUI

/// The bar under the grid while edits are staged.
struct StagedEditsBar: View {
    @ObservedObject var model: AppModel
    @ObservedObject var edits: PendingEdits

    var body: some View {
        HStack(spacing: 10) {
            Image(systemName: "square.and.pencil")
                .foregroundStyle(Color.accentColor)
                .imageScale(.medium)

            VStack(alignment: .leading, spacing: 1) {
                Text(headline)
                    .font(.system(size: 12, weight: .semibold))
                if let detail {
                    Text(detail)
                        .font(.system(size: 10.5))
                        .foregroundStyle(.secondary)
                }
            }

            Spacer(minLength: 8)

            if model.isCommitting {
                ProgressView()
                    .controlSize(.small)
                Text("committing…")
                    .font(.system(size: 11))
                    .foregroundStyle(.secondary)
            } else if edits.pendingCount == 0 {
                Button("Reload") { model.reloadResult() }
                    .buttonStyle(.borderedProminent)
                    .controlSize(.small)
            } else {
                Button("Discard") { model.discardStagedEdits() }
                    .buttonStyle(.bordered)
                    .controlSize(.small)
                if edits.conflictCount > 0 {
                    Button(conflictTitle) { model.reviewConflicts() }
                        .buttonStyle(.borderedProminent)
                        .controlSize(.small)
                        .tint(.orange)
                        .disabled(model.isRereading)
                }
                Button(commitTitle) { model.commitStagedEdits() }
                    .buttonStyle(.borderedProminent)
                    .controlSize(.small)
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 7)
        .background(.ultraThinMaterial)
        .overlay(alignment: .top) {
            Rectangle().fill(Color.accentColor.opacity(0.35)).frame(height: 1)
        }
    }

    private var headline: String {
        let n = edits.pendingCount
        if n == 0 {
            let written = edits.documents.count
            return written == 1
                ? "1 document written — the grid still shows what was loaded"
                : "\(written) documents written — the grid still shows what was loaded"
        }
        return n == 1
            ? "1 document edited, not yet written"
            : "\(n) documents edited, not yet written"
    }

    private var detail: String? {
        var parts: [String] = []
        if edits.updateCount > 0 { parts.append("\(edits.updateCount) to update") }
        if edits.deleteCount > 0 { parts.append("\(edits.deleteCount) to delete") }
        let applied = edits.documents.count - edits.pendingCount
        if applied > 0 { parts.append("\(applied) already written") }
        if edits.conflictCount > 0 {
            parts.append(
                edits.conflictCount == 1
                    ? "1 changed on the server" : "\(edits.conflictCount) changed on the server")
        }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    private var commitTitle: String {
        edits.pendingCount == 1 ? "Commit 1…" : "Commit \(edits.pendingCount)…"
    }

    private var conflictTitle: String {
        edits.conflictCount == 1 ? "Resolve 1 Conflict…" : "Resolve \(edits.conflictCount) Conflicts…"
    }
}

// MARK: - the report

/// What the commit actually did, per document.
struct MutationReportSheet: View {
    @ObservedObject var model: AppModel
    let report: MutationReport

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            Divider()
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    if !report.notices.isEmpty { notices }
                    ForEach(report.rows) { row in
                        MutationRowLine(row: row)
                    }
                }
                .padding(14)
            }
            Divider()
            footer
        }
        .frame(width: 560, height: 420)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(title)
                .font(.system(size: 14, weight: .semibold))
            Text(subtitle)
                .font(.system(size: 11.5))
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(14)
    }

    private var title: String {
        if report.isClean {
            return report.applied == 1
                ? "1 document written" : "\(report.applied) documents written"
        }
        return "The batch stopped part way through"
    }

    private var subtitle: String {
        var parts = ["\(report.applied) applied"]
        if report.failed > 0 { parts.append("\(report.failed) failed") }
        if report.notAttempted > 0 { parts.append("\(report.notAttempted) never attempted") }
        var line = parts.joined(separator: " · ")
        if report.notAttempted > 0 {
            line +=
                ". The ones that were never attempted are still staged — nothing was written for them, and nothing was lost."
        }
        if report.conflicts > 0 {
            line +=
                " A version conflict means the document changed on the server after you loaded it, so the write was refused rather than overwriting someone else's change. What you typed is still staged — resolve it below to see what changed."
        }
        return line
    }

    private var notices: some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(report.notices) { notice in
                HStack(alignment: .firstTextBaseline, spacing: 7) {
                    Image(
                        systemName: notice.isWarning
                            ? "exclamationmark.triangle.fill" : "info.circle.fill"
                    )
                    .foregroundStyle(notice.isWarning ? Color.orange : Color.secondary)
                    .imageScale(.small)
                    VStack(alignment: .leading, spacing: 1) {
                        Text(notice.message)
                            .font(.system(size: 11.5))
                            .fixedSize(horizontal: false, vertical: true)
                        if let code = notice.code {
                            Text(code)
                                .font(.system(size: 10, design: .monospaced))
                                .foregroundStyle(.tertiary)
                        }
                    }
                }
            }
            Divider().padding(.vertical, 2)
        }
    }

    private var footer: some View {
        HStack {
            Text(footerNote)
                .font(.system(size: 11))
                .foregroundStyle(.secondary)
            Spacer()
            if report.conflicts > 0 {
                Button("Resolve Conflicts…") { model.reviewConflicts() }
                    .disabled(model.isRereading)
            }
            Button("Done") { model.showMutationReport = false }
                .keyboardShortcut(.defaultAction)
        }
        .padding(12)
    }

    private var footerNote: String {
        if report.conflicts > 0 {
            return "Reads each conflicted document back and shows what changed."
        }
        return report.isClean ? "" : "Re-run the statement to see what the server holds now."
    }
}

/// One document's line in the report.
private struct MutationRowLine: View {
    let row: MutationRow

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 9) {
            Image(systemName: symbol)
                .foregroundStyle(tint)
                .imageScale(.small)
                .frame(width: 14)
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    Text(row.op)
                        .font(.system(size: 10, weight: .semibold))
                        .foregroundStyle(.secondary)
                    Text("\(row.index)/\(row.documentID)")
                        .font(.system(size: 11.5, design: .monospaced))
                }
                if let detail {
                    Text(detail)
                        .font(.system(size: 11))
                        .foregroundStyle(row.outcome == .applied ? .secondary : .primary)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
            Spacer(minLength: 6)
        }
    }

    private var symbol: String {
        switch row.outcome {
        case .applied: return "checkmark.circle.fill"
        case .failed: return row.conflict ? "arrow.triangle.branch" : "xmark.circle.fill"
        case .notAttempted: return "clock"
        }
    }

    private var tint: Color {
        switch row.outcome {
        case .applied: return .green
        case .failed: return row.conflict ? .orange : .red
        case .notAttempted: return .secondary
        }
    }

    private var detail: String? {
        switch row.outcome {
        case .applied:
            var line = "written"
            if let seq = row.seqNo { line += " · now at _seq_no \(seq)" }
            if row.forcedRefresh {
                line += " · the server forced an immediate refresh rather than waiting for one"
            }
            return line
        case .notAttempted:
            return "never attempted — the batch stopped before it, so this is still staged"
        case .failed:
            if row.conflict {
                return
                    "version conflict — this document changed on the server after you loaded it, so nothing was written"
            }
            return row.error ?? "the write failed"
        }
    }
}
